File: python/bigasterisk/query.py
"""What the tools re-run.

Several of the tools here are search procedures: fuzzing, symbolic test generation and
delta debugging all work by proposing an input, running the query with it, and looking
at what came out — dozens or hundreds of times over.

That makes "the query" not a result but a recipe, and a recipe can arrive in two forms.
A SQL string re-runs by being re-parsed, so whatever is registered under ``flights`` at
the time is what it reads. A DataFrame cannot: by the time you hold one, its plan is
analysed and bound, and ``flights`` has stopped being a name. Running it again runs it
against the same data.

Requiring the string form would mean rewriting a PySpark pipeline as SQL before you
could debug it — which is not the program you set out to debug. So the tools take
either, and the JVM side substitutes into a DataFrame's plan when it has to. That is
what :func:`as_query` hands across the boundary.
"""


def as_query(query):
    """``query`` in the form the JVM tools expect.

    A string goes over as a string; a DataFrame goes over as the underlying Java
    ``Dataset``. Anything else is refused here rather than becoming a Py4J error about
    an overload that does not exist.
    """
    if isinstance(query, str):
        return query
    jdf = getattr(query, "_jdf", None)
    if jdf is not None:
        return jdf
    raise TypeError(
        "a query must be SQL text or a DataFrame, not %s" % type(query).__name__)

File: modules/api/src/main/scala/org/bigasterisk/api/Query.scala
package org.bigasterisk.api

import scala.language.implicitConversions

import org.apache.spark.sql.{DataFrame, Dataset}

/**
 * Something the tools can run more than once, against whatever its input tables hold
 * at the time.
 *
 * Several of the tools here are search procedures: they substitute data under the
 * query's tables and run it again, dozens or hundreds of times. Fuzzing generates the
 * substitute; test generation solves for it; delta debugging removes rows from it. What
 * they all need is not a result but a *recipe* — something that will read whatever is
 * under `flights` the next time it is asked.
 *
 * SQL text is such a recipe already: re-parsing it picks up whatever the name now
 * resolves to. A DataFrame is not. By the time you hold one, its plan is analysed and
 * bound: `flights` has become a particular relation with particular attribute ids, and
 * running it again runs it against the same data. That is why the tools originally took
 * the query as text.
 *
 * Requiring text is the wrong trade. It says that to debug a DataFrame pipeline you
 * must first rewrite it as a SQL string — which is not the program you are debugging.
 * So a [[Query]] is either, and the binding is responsible for making a DataFrame
 * re-runnable by substituting into its plan (see the binding's plan rebinding: seed
 * plans are matched inside the query's plan and replaced, with the original output
 * attributes preserved so everything above them still refers to the right columns).
 *
 * Both forms convert implicitly, so `fuzz("SELECT ...", seeds)` and `fuzz(df, seeds)`
 * are both just calls.
 *
 * @group entry
 */
sealed trait Query {

  /** How to describe this query in a message, briefly. */
  def describe: String
}

/** @group entry */
object Query {

  /** A query given as SQL text, run by parsing it against the current catalog. */
  final case class Sql(text: String) extends Query {
    override def describe: String = text.replaceAll("\\s+", " ").trim
  }

  /**
   * A query given as a DataFrame, run by substituting into its analysed plan.
   *
   * @param df the pipeline, built however you build pipelines
   */
  final case class Frame(df: DataFrame) extends Query {
    override def describe: String = df.queryExecution.analyzed.nodeName + " ..."
  }

  implicit def fromText(text: String): Query = Sql(text)

  implicit def fromFrame(df: DataFrame): Query = Frame(df)

  /** The query as text, if it was given as text. */
  def textOf(query: Query): Option[String] = query match {
    case Sql(text) => Some(text)
    case _ => None
  }

  /**
   * A [[Query]] from whatever a foreign caller handed over.
   *
   * The implicit conversions above are a compile-time convenience and so are no use to
   * Java or to Py4J, which arrives holding a `String` or a `Dataset` and no static type
   * at all. This is the same choice made at run time.
   */
  def of(query: AnyRef): Query = query match {
    case q: Query => q
    case text: String => Sql(text)
    // `DataFrame` is `Dataset[Row]`, and erasure leaves no `Row` to test for; a
    // `Dataset` of anything else would fail later, and more obscurely, than here.
    case ds: Dataset[_] => Frame(ds.toDF())
    case other =>
      throw new IllegalArgumentException(
        "a query must be SQL text or a DataFrame, not " +
        (if (other == null) "null" else other.getClass.getName))
  }
}
